app/javascript/controllers/creator_settings_controller.js

```diff
@@ -0,0 +1,57 @@
+import { Controller } from '@hotwired/stimulus';
+
+const MAX_LOGO_PREVIEW_HEIGHT = 80;
+const MAX_LOGO_PREVIEW_WIDTH = 220;
+
+export class CreatorSettingsController extends Controller {
+  static targets = ['previewLogo'];
+
+  previewLogo(event) {
+    const {
+      target: {
+        files: [firstFile],
+      },
+    } = event;
+
+    const reader = new FileReader();
+
+    reader.onload = () => {
+      const imageURL = reader.result;
+      const image = document.createElement('img');
+      image.src = imageURL;
+
+      // The logo preview image is purely visual so no need to communicate this to assistive technology.
+      image.alt = 'preview of logo selected';
+
+      image.addEventListener(
+        'load',
+        (event) => {
+          let {
+            target: { width, height },
+          } = event;
+
+          if (height > MAX_LOGO_PREVIEW_HEIGHT) {
+            width = (width / height) * MAX_LOGO_PREVIEW_HEIGHT;
+            height = MAX_LOGO_PREVIEW_HEIGHT;
+          }
+
+          if (width > MAX_LOGO_PREVIEW_WIDTH) {
+            width = MAX_LOGO_PREVIEW_WIDTH;
+            height = (width / height) * MAX_LOGO_PREVIEW_WIDTH;
+          }
+
+          image.style.width = `${width}px`;
+          image.style.height = `${height}px`;
+
+          this.previewLogoTarget.replaceChild(
+            image,
+            this.previewLogoTarget.firstChild,
+          );
+        },
+        { once: true },
+      );
+    };
+
+    reader.readAsDataURL(firstFile);
```

Comment:
I'm seeing an error in the console in the case where a user opens the file picker, but instead of choosing a file, clicks "Cancel":
![error executing 'readDataAsURL'](https://user-images.githubusercontent.com/20773163/141983315-82166016-f77f-434b-887c-72868549c098.png)

I think that made me get into this weird state where I had:

- picked an image, and saw the preview as expected
- opened the image picker again, and clicked cancel
- couldn't submit the form, but no error message was visible

https://user-images.githubusercontent.com/20773163/141983453-af40b468-5624-4791-ad45-0d8eb8200427.mp4

Reply:
Good ctach. I hadn't considered the cancel. I'll get that sorted.